src/datahike/db.cljc

```diff
@@ -1375,277 +1495,339 @@
-            (if-let [vid (get tempids v)]
-              (recur report (cons [op e a vid] entities))
-              (recur (allocate-eid report v (next-eid db)) es))
-
-            (= op :db/add)
-            (recur (transact-add report entity) entities)
-
-            (= op :db/retract)
-            (if-some [e (entid db e)]
-              (let [v (if (ref? db a) (entid-strict db v) v)]
-                (validate-attr a entity db)
-                (validate-val v entity db)
-                (if-some [old-datom (first (-search db [e a v]))]
-                  (recur (transact-retract-datom report old-datom) entities)
-                  (recur report
-                         entities)))
-              (recur report entities))
-
-            (= op :db.fn/retractAttribute)
-            (if-let [e (entid db e)]
-              (let [_ (validate-attr a entity db)
-                    datoms (vec (-search db [e a]))]
-                (recur (reduce transact-retract-datom report datoms)
-                       (concat (retract-components db datoms) entities)))
-              (recur report entities))
-
-            (or (= op :db.fn/retractEntity)
+                        (raise ":db.fn/cas failed on datom [" e " " a " " (map :v datoms) "], expected " ov
+                          {:error :transact/cas, :old datoms, :expected ov, :new nv}))
+                      (let [v (:v (first datoms))]
+                        (if (= v ov)
+                          (recur (transact-add report [:db/add e a nv]) entities)
+                          (raise ":db.fn/cas failed on datom [" e " " a " " v "], expected " ov
+                            {:error :transact/cas, :old (first datoms), :expected ov, :new nv})))))))
+
+              (tx-id? e)
+              (recur (allocate-eid report e (current-tx report)) (cons [op (current-tx report) a v] entities))
+
+              (and (ref? db a) (tx-id? v))
+              (recur (allocate-eid report v (current-tx report)) (cons [op e a (current-tx report)] entities))
+
+              (tempid? e)
+              (let [upserted-eid (when (is-attr? db a :db.unique/identity)
+                                   (:e (first (-datoms db :avet [a v]))))
+                    allocated-eid (get tempids e)]
+                (if (and upserted-eid allocated-eid (not= upserted-eid allocated-eid))
+                  (retry-with-tempid initial-report report initial-es e upserted-eid)
+                  (let [eid (or upserted-eid allocated-eid (next-eid db))]
+                    (recur (allocate-eid report e eid) (cons [op eid a v] entities)))))
+
+              (and (ref? db a) (tempid? v))
+              (if-let [vid (get tempids v)]
+                (recur report (cons [op e a vid] entities))
+                (recur (allocate-eid report v (next-eid db)) es))
+
+              (and (homogeneous-tuple-attr? db a)
+                (> (count v) 8))
+              (raise "Cannot store more than 8 values for homogeneous tuple: " entity
```

Comment:
Probably it suffices to send the corresponding entity in the attached map?

Reply:
but on the other hand it is consistent with some other raises. I don't know... I would probably leave it to the map but you can decide.